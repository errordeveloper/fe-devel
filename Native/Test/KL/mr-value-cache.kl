/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

operator gen(io String output) {
  report("  Running generator!");
  output = "Hello";
}

operator testCachedGenerator()
{
  report("--- Testing Cached Generator");
  ValueProducer<String> vp = createValueCache(
    createValueGenerator( gen )
  );
  report("vp = " + vp);

  report("Should run generator");
  report(" vp.produce() = " + vp.produce());

  report("Should not run generator (use cache)");
  report(" vp.produce() = " + vp.produce());

  vp.flush();
  report("Should run generator");
  report(" vp.produce() = " + vp.produce());
  report(" ");
}

operator multByTwo(
  io Integer value
  )
{
  report("  Running transform!");
  value *= 2;
}

operator testCachedTransform()
{
  report("--- Testing Cached Transform");
  ValueProducer<Integer> vp = createValueCache(
    createValueTransform(
      createConstValue( 42 ),
      multByTwo
      )
    );
  report("vp = " + vp);

  report("Should run transform");
  report(" vp.produce() = " + vp.produce());

  report("Should not run transform (use cache)");
  report(" vp.produce() = " + vp.produce());

  vp.flush();
  report("Should run transform");
  report(" vp.produce() = " + vp.produce());
  report(" ");
}

operator testNestedCache()
{
  report("--- Testing Nested Caches");
  ValueProducer<String> vp = createValueCache(
    createValueCache(
      createValueGenerator( gen )
    )
  );
  report("vp = " + vp);

  report("Should run generator");
  report(" vp.produce() = " + vp.produce());

  report("Should not run generator (use cache)");
  report(" vp.produce() = " + vp.produce());

  vp.flush();
  report("Should run generator");
  report(" vp.produce() = " + vp.produce());
  report(" ");
}

operator genInteger(io Integer output) {
  report("  Running generator!");
  output = 123;
}

operator testTransformFlushPropagate()
{
  report("--- Testing Transform flush propagation");
  ValueProducer<Integer> vp = 
    createValueTransform(
      createValueCache( createValueGenerator( genInteger ) ),
      multByTwo
    );
  report("vp = " + vp);

  report("Should run generator");
  report(" vp.produce() = " + vp.produce());

  report("Should not run generator (use cache)");
  report(" vp.produce() = " + vp.produce());

  vp.flush();
  report("Should run generator");
  report(" vp.produce() = " + vp.produce());
  report(" ");
}

operator multByPi(
  Integer input,
  io Scalar output
  )
{
  output = 3.14 * input;
}

operator testMapFlushPropagate()
{
  report("--- Testing Map flush propagation");
  ValueProducer<Scalar> vp = 
    createValueMap(
      createValueCache( createValueGenerator( genInteger ) ),
      multByPi
    );
  report("vp = " + vp);

  report("Should run generator");
  report(" vp.produce() = " + vp.produce());

  report("Should not run generator (use cache)");
  report(" vp.produce() = " + vp.produce());

  vp.flush();
  report("Should run generator");
  report(" vp.produce() = " + vp.produce());
  report(" ");
}

operator entry() {
  testCachedGenerator();
  testCachedTransform();
  testNestedCache();
  testTransformFlushPropagate();
  testMapFlushPropagate();
}

