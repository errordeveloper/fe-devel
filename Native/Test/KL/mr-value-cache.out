--- Testing Cached Generator
vp = ValueProducer<String>
Should run generator
  Running generator!
 vp.produce() = Hello
Should not run generator (use cache)
 vp.produce() = Hello
Should run generator
  Running generator!
 vp.produce() = Hello
 
--- Testing Cached Transform
vp = ValueProducer<Integer>
Should run transform
  Running transform!
 vp.produce() = 84
Should not run transform (use cache)
 vp.produce() = 84
Should run transform
  Running transform!
 vp.produce() = 84
 
--- Testing Nested Caches
vp = ValueProducer<String>
Should run generator
  Running generator!
 vp.produce() = Hello
Should not run generator (use cache)
 vp.produce() = Hello
Should run generator
  Running generator!
 vp.produce() = Hello
 
--- Testing Transform flush propagation
vp = ValueProducer<Integer>
Should run generator
  Running generator!
  Running transform!
 vp.produce() = 246
Should not run generator (use cache)
  Running transform!
 vp.produce() = 246
Should run generator
  Running generator!
  Running transform!
 vp.produce() = 246
 
--- Testing Map flush propagation
vp = ValueProducer<Scalar>
Should run generator
  Running generator!
 vp.produce() = 386.22
Should not run generator (use cache)
 vp.produce() = 386.22
Should run generator
  Running generator!
 vp.produce() = 386.22
 
